Nr=(17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36)
for ((i=0; i<${#Nr[@]}; i++)); do cd ~/fastq_files_splitted/ && prefetch SRR197604${Nr[i]} && fastq-dump --split-3 SRR197604${Nr[i]} && cd .. && conda activate trimming && trim_galore --illumina --paired --fastqc -o trim_galore ~/fastq_files_splitted/SRR197604${Nr[i]}_1.fastq ~/fastq_files_splitted/SRR197604${Nr[i]}_2.fastq && conda deactivate && conda activate mapping && STAR --outSAMtype BAM Unsorted --genomeDir ~/RefSeq/ref --runThreadN 5 --readFilesIn ~/trim_galore/SRR197604${Nr[i]}_1_val_1.fq ~/trim_galore/SRR197604${Nr[i]}_2_val_2.fq --outFileNamePrefix ~/STAR/SRR197604${Nr[i]} && conda deactivate && conda activate extract && rm ~/fastq_files_splitted/SRR197604${Nr[i]}*.fastq && rm ~/trim_galore/SRR197604${Nr[i]}*.fq && rm -r ~/fastq_files_splitted/SRR197604${Nr[i]} && samtools sort ~/STAR/SRR197604${Nr[i]}Aligned.out.bam -o ~/sorted/sorted_SRR197604${Nr[i]}.bam && conda deactivate && rm -r ~/STAR/SRR197604${Nr[i]}* && featureCounts -p -O -T 4 -a ~/RefSeq/genomic.gtf -o ~/Counts/SRR197604${Nr[i]}_Counts.txt ~/sorted/sorted_SRR197604${Nr[i]}.bam; done